cli/azd/pkg/extensions/source.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package extensions
+
+import (
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"log"
+	"path"
+	"slices"
+	"strings"
+
+	"github.com/azure/azure-dev/cli/azd/resources"
+)
+
+type Source interface {
+	// Name returns the name of the source.
+	Name() string
+	// ListTemplates returns a list of AZD compatible templates.
+	ListExtensions(ctx context.Context) ([]*ExtensionMetadata, error)
+	// GetTemplate returns a template by path.
+	GetExtension(ctx context.Context, name string) (*ExtensionMetadata, error)
+}
+
+type registrySource struct {
+	name     string
+	registry *Registry
+}
+
+// newRegistrySource creates a new registry source.
+func newRegistrySource(name string, registry *Registry) (Source, error) {
+	if err := validateRegistry(*registry); err != nil {
+		return nil, fmt.Errorf("failed to validate registry: %w", err)
+	}
+
+	return &registrySource{
+		name:     name,
+		registry: registry,
+	}, nil
+}
+
+func (ts *registrySource) Name() string {
+	return ts.name
+}
+
+// ListTemplates returns a list of templates from the extension source.
+func (s *registrySource) ListExtensions(ctx context.Context) ([]*ExtensionMetadata, error) {
+	for _, extension := range s.registry.Extensions {
+		extension.Source = s.name
+	}
+
+	return s.registry.Extensions, nil
+}
+
+// GetExtension returns an extension by id.
+func (s *registrySource) GetExtension(ctx context.Context, id string) (*ExtensionMetadata, error) {
+	allTemplates, err := s.ListExtensions(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed listing templates: %w", err)
+	}
+
+	matchingIndex := slices.IndexFunc(allTemplates, func(extension *ExtensionMetadata) bool {
+		return strings.EqualFold(extension.Id, id)
+	})
+
+	if matchingIndex == -1 {
+		return nil, fmt.Errorf("'%s' %w", id, ErrRegistryExtensionNotFound)
+	}
+
+	return allTemplates[matchingIndex], nil
+}
+
+// validateRegistry validates the registry content and its signature
+func validateRegistry(registry Registry) error {
+	if registry.Signature == "" {
```

Comment:
Just confirming -- have we managed to run this through a security review, as previously suggested in [4409](https://github.com/Azure/azure-dev/pull/4409#pullrequestreview-2452796364) by Matt?

I think this means that if no Signature is included in the hosted registry, we should bypass the signature validation. Supposedly, this would open up MITM spoofing that we are intending to prevent